src/peerly/services/p2pPhoneListUpload.service.ts

```diff
@@ -0,0 +1,143 @@
+import { CampaignTcrComplianceService } from '../../campaigns/tcrCompliance/services/campaignTcrCompliance.service'
+import { P2pPhoneListRequestSchema } from '../schemas/p2pPhoneListRequest.schema'
+import {
+  VoterFileType,
+  CustomFilter,
+} from '../../voters/voterFile/voterFile.types'
+import { typeToQuery } from '../../voters/voterFile/util/voterFile.util'
+import { Readable } from 'stream'
+import { to as copyTo } from 'pg-copy-streams'
+
+@Injectable()
+export class P2pPhoneListUploadService {
+  private readonly logger = new Logger(P2pPhoneListUploadService.name)
+
+  constructor(
+    private readonly voterFileFilterService: VoterFileFilterService,
+    private readonly voterDatabaseService: VoterDatabaseService,
+    private readonly peerlyPhoneListService: PeerlyPhoneListService,
+    private readonly tcrComplianceService: CampaignTcrComplianceService,
+  ) {}
+
+  async uploadPhoneList(
+    campaign: Campaign,
+    request: P2pPhoneListRequestSchema,
+  ): Promise<{ token: string; listName: string }> {
+    const { listName, ...filterData } = request
+
+    // Get TCR compliance record for identity_id
+    const tcrCompliance = await this.tcrComplianceService.fetchByCampaignId(
+      campaign.id,
+    )
+
+    if (!tcrCompliance) {
+      throw new NotFoundException(
+        'TCR compliance record not found for this campaign',
+      )
+    }
+
+    if (!tcrCompliance.peerlyIdentityId) {
+      throw new BadRequestException(
+        'TCR compliance record does not have a Peerly identity ID',
+      )
+    }
+
+    // Create a temporary voter file filter to generate the audience data
+    const tempFilterName = `P2P_${listName}_${Date.now()}`
+    const voterFileFilter = await this.voterFileFilterService.create(
+      campaign.id,
+      {
+        name: tempFilterName,
+        ...filterData,
+      },
+    )
+
+    try {
+      // Convert voter file filter to audience format for CSV generation
+      const audienceData =
+        await this.voterFileFilterService.voterFileFilterToAudience(
+          voterFileFilter,
+        )
+
+      // Convert audience data to filters array
+      const filters: CustomFilter[] = Object.entries(audienceData)
+        .filter(([_key, value]) => value === true)
+        .map(([key, _value]) => key as CustomFilter)
+
+      // Generate CSV stream with voter data including phone numbers
+      const csvStream = await this.generatePhoneListCsvStream(campaign, filters)
+
+      // Upload to Peerly
+      const token = await this.peerlyPhoneListService.uploadPhoneListToken({
+        listName,
+        csvStream,
+        identityId: tcrCompliance.peerlyIdentityId,
+      })
+
+      this.logger.log(
+        `P2P phone list uploaded successfully for campaign ${campaign.id}, token: ${token}`,
+      )
+
+      return { token, listName }
+    } finally {
+      // Clean up temporary voter file filter
+      await this.voterFileFilterService.delete(voterFileFilter.id)
+    }
```

Comment:
Whatever we end up doing here, we most def don't want to be writing a "temporary" record to the DB, that very likely won't be cleaned up in the case of error handlings.

wrapping whole chunks of code like this in `try-catch`es is also an anti-pattern and complicates error/exception handling. `try-catch`es should be targeted around succinct lines of code that we know will `throw` if an error/exception will be generated (i.e. `fetch` calls)